### PR TITLE
network: Wait for network device in UpdateInterface

### LINKED ISSUE #305: UpdateInterface should now check if the network device is available before configuring the interface.
Since we plan to support hotplugging network devices, UpdateInterface should wait on the hotplugged device based on its PCI address and then configure the device.

### PR DESCRIPTION
**Pull request by amshinde**
A new field PciAddr has been added to grpc.Interface.
This is the PCI identifier of the network device.

In order to support network hotplug, we need to wait for the
device to be online first based on the PCI address of the device
inside the VM before updating its configuration.

Fixes #305

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

### CODE REVIEW

File: protocols/grpc/agent.proto [hunk 1]
```diff
@@ -279,6 +279,11 @@ message Interface {
 	repeated IPAddress IPAddresses = 3;
 	uint64 mtu = 4;
 	string hwAddr = 5;
+
+	// pciAddr should be the PCI address in the format  "bridgeAddr/deviceAddr".
```
Comment (jodh-intel) [inline protocols/grpc/agent.proto]: s/should be/is/ ?

File: protocols/grpc/agent.proto [hunk 2]
```diff
@@ -279,6 +279,11 @@ message Interface {
...
+	// Here, bridgeAddr is the address at which the brige is attached on the root bus,
```
Comment (jodh-intel) [inline protocols/grpc/agent.proto]: s/brige/bridge/

**Comment by amshinde** [comment]
@sboeuf I have introduced a few changes for this PR. Can you try out this branch here: https://github.com/amshinde/agent-1/tree/wait-network-interface2

**Comment by sboeuf** [comment]
@amshinde I'll check that, thx !

**Comment by sboeuf** [comment]
@amshinde I've just tried it and it does not work. Seems like the interface never gets detected...
Have you tried your code? It seems sane to me to wait for a uevent about interface, but apparently we're not getting it.

**Comment by amshinde** [comment]
@sboeuf I had tested with qmp hotplugging a network device. Can you check if you even get the uevent in your case, you can add a log here : https://github.com/kata-containers/agent/blob/master/agent.go#L494.

Let me know if are able to try it out, else I will give a shot setting up the watcher itself later today.

**Comment by amshinde** [comment]
@sboeuf I have tested with your branch, taking out the sleep that you had added and verified that this works.

File: network.go [hunk 3]
```diff
@@ -239,8 +242,36 @@ func (s *sandbox) updateInterface(netHandle *netlink.Handle, iface *pb.Interface
...
+		"pci-address":    iface.PciAddr,
 	})
 
+	// If the PCI address of the network device is provided, wait/check for the device
+	// to be available first
+
+	if iface.PciAddr != "" {
+		// iface.PciAddr is in the format bridgeAddr/deviceAddr eg. 05/06
+		// Get the actual PCI address inside the VM in the format "0000:01:06.0"
+		pciAddr, err := getDevicePCIAddress(iface.PciAddr)
```
[the remark this answers is not in the document]
Comment (amshinde) [reply network.go]: @bergwolf I have reworked this PR a bit. Can you take a look?

**Comment by amshinde** [comment]
@jodh-intel @bergwolf @sboeuf PTAL

**Comment by raravena80** [comment]
Ping @amshinde. Ready to merge? There are some conflicts

**Comment by amshinde** [comment]
Rebased and pushed.

**Pull request merged**